File: lintcode/convert_sorted_array_to_BST.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param A: a list of integer
    @return: a tree node
    """
    def sortedArrayToBST(self, A):
        def helper(A, begin, end):
            if begin == end: return None
            mid = begin + (end - begin) / 2
            root = TreeNode(A[mid])
            root.left, root.right = helper(A, begin, mid), \
                                    helper(A, mid + 1, end)
            return root
        return helper(A, 0, len(A))

File: leetcode/convert_sorted_array_to_BST.js
/* Given an array where elements are sorted in ascending order, 
 * convert it to a height balanced BST.
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function(nums) {
    function helper(nums, begin, end) {
        if (begin === end)
            return null
        var mid = begin + Math.floor((end - begin) / 2)
        var root = new TreeNode(nums[mid])
        root.left = helper(nums, begin, mid)
        root.right = helper(nums, mid + 1, end)
        return root
    }
    return helper(nums, 0, nums.length)
};
